### PR TITLE
agent self-update

### PR DESCRIPTION
**Pull request by TingluoHuang**

### CODE REVIEW

**Comment by bryanmacfarlane** [comment]
This ready to go in?

**Comment by TingluoHuang** [comment]
@bryanmacfarlane haven't. waiting for Scott's webapi changes.

File: src/Agent.Listener/SelfUpdater.cs [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+    }
+
+    public class SelfUpdater : AgentService, ISelfUpdater
+    {
+#if OS_WINDOWS
+        private static string _packageName = "agentcore-windows";
+#elif OS_OSX
+        private static string _packageName = "agentcore-osx";
+#elif OS_LINUX
+        private static string _packageName = "agentcore-linuex";
```
Comment (bryanmacfarlane) [inline src/Agent.Listener/SelfUpdater.cs]: is linuex intended?  extra e?

Comment (TingluoHuang) [inline src/Agent.Listener/SelfUpdater.cs]: it's not. :D will remove it.

File: src/Agent.Listener/Agent.cs [hunk 2]
```diff
@@ -177,20 +182,72 @@ private void CtrlCHandler(object sender, EventArgs e)
...
+                                    {
+                                        autoUpdateInProgress = true;
+                                        // TODO: autoupdate
+                                        var selfUpdater = HostContext.GetService<ISelfUpdater>();
+                                        selfUpdateTask = selfUpdater.SelfUpdate(workerManager, !settings.RunAsService, token);
+                                        Trace.Warning("Referesh message received, kick-off selfupdate background process.");
+                                    }
+                                    else
+                                    {
+                                        Trace.Info("Referesh message received, skip autoupdate since a previous autoupdate is already running.");
```
Comment (ericsciple) [inline src/Agent.Listener/Agent.cs]: refresh!!!

File: src/Agent.Listener/Agent.cs [hunk 3]
```diff
@@ -177,20 +182,72 @@ private void CtrlCHandler(object sender, EventArgs e)
...
+                                }
                             }
                             else if (string.Equals(message.MessageType, JobRequestMessage.MessageType, StringComparison.OrdinalIgnoreCase))
                             {
-                                var newJobMessage = JsonUtility.FromString<JobRequestMessage>(message.Body);
-                                workerManager.Run(newJobMessage);
+                                if (autoUpdateInProgress)
+                                {
+                                    skipMessageDeletion = true;
+                                    break;
```
Comment (ericsciple) [inline src/Agent.Listener/Agent.cs]: shouldn't break.

File: src/Agent.Listener/Agent.cs [hunk 4]
```diff
@@ -177,20 +182,72 @@ private void CtrlCHandler(object sender, EventArgs e)
...
+                                if (completeTask == getNextMessage)
+                                {
+                                    message = await getNextMessage;
+                                }
+                                else
+                                {
+                                    autoUpdateInProgress = false;
+                                    if (await selfUpdateTask)
+                                    {
+                                        return 2;
```
Comment (ericsciple) [inline src/Agent.Listener/Agent.cs]: change to 3,

File: src/Agent.Listener/Agent.cs [hunk 5]
```diff
@@ -177,20 +182,72 @@ private void CtrlCHandler(object sender, EventArgs e)
...
                     while (!token.IsCancellationRequested)
                     {
                         try
                         {
-                            message = await listener.GetNextMessageAsync(token); //get next message
-
+                            if (autoUpdateInProgress)
+                            {
+                                Task<TaskAgentMessage> getNextMessage = listener.GetNextMessageAsync(token);
+                                Task completeTask= await Task.WhenAny(getNextMessage, selfUpdateTask);
```
Comment (ericsciple) [inline src/Agent.Listener/Agent.cs]: format:D

**Comment by ericsciple** [comment]
GREAT!

**Comment by bryanmacfarlane** [comment]
Awesome!

**Pull request merged**